### PR TITLE
Develop console rewrite

### PR DESCRIPTION
**Pull request by bendikro**
A rewrite of most of the console code. Should work for the most part, but bugs probably exist.

Still a bit in development:
- [X] Unresolved issue with search in torrent list

Fixes the following tickets:
- [#1119](http://dev.deluge-torrent.org/ticket/1119): logging doesn't work with deluge-console because of ncurses
- [#2097](http://dev.deluge-torrent.org/ticket/2097): Corrected spelling error and added --intense_match to console(legacy and non-interactive)
- [#2011](http://dev.deluge-torrent.org/ticket/2011): deluge-console: counterintuitive behavior of rm
- [#2470](http://dev.deluge-torrent.org/ticket/2470): deluge-console's parsing of commands is surprising
- [#2099](http://dev.deluge-torrent.org/ticket/2099): console does not support monochrome terminals
- [#2838](http://dev.deluge-torrent.org/ticket/2838): missing ! in ui/console/colors.py line 59
- [#2914](http://dev.deluge-torrent.org/ticket/2914): Config migration, file format and default config

The following pylint warnings can be removed on this branch with no errors:
`super-on-old-class, pointless-except, non-parent-init-called`

`super-init-not-called` has only a couple of warnings in `deluge/error.py` and `deluge/tests/test_httpdownloader.py`

### CODE REVIEW

File: deluge/log.py [hunk 1]
```diff
@@ -107,7 +107,8 @@ def findCaller(self):  # NOQA
 }
 
 
-def setup_logger(level="error", filename=None, filemode="w", logrotate=None):
+def setup_logger(level="error", filename=None, filemode="w", logrotate=None,
+                 default_output_stream=sys.stdout):
```
Comment (cas--) [inline deluge/log.py]: Should it just be 'output_stream'?

**Comment by cas--** [comment]
How do you feel about replacing `optparse` for `argsparse` in `console/main.py`? :)

**Comment by cas--** [comment]
A different question instead... could it perhaps be confusing using `argparse` and `OptionsParser` is the class in console?

**Comment by bendikro** [comment]
Yeah, can fix that

**Comment by cas--** [comment]
Added markup to help strings

File: deluge/ui/console/commands/config.py [hunk 2]
```diff
@@ -65,14 +65,14 @@ def simple_eval(source):
 class Command(BaseCommand):
     """Show and set configuration values"""
 
-    usage = "config [--set <key> <value>] [<key> [<key>...] ]"
+    usage = _("Usage: config [--set <key> <value>] [<key> [<key>...] ]")
```
Comment (bendikro) [inline deluge/ui/console/commands/config.py]: I wonder whether it's a good idea to translate strings like this which includes argument option description.
If `key` and `value` should be translated, they must also be translated in the other arguments below.

File: deluge/ui/console/commands/config.py [hunk 3]
```diff
@@ -65,14 +65,14 @@ def simple_eval(source):
...
 
     def add_arguments(self, parser):
         set_group = parser.add_argument_group("setting a value")
-        set_group.add_argument("-s", "--set", action="store", metavar="<key>", help="set value for this key")
-        set_group.add_argument("values", metavar="<value>", nargs="+", help="Value to set")
+        set_group.add_argument("-s", "--set", action="store", metavar="<key>", help=_("set value for this key"))
+        set_group.add_argument("values", metavar="<value>", nargs="+", help=_("Value to set"))
```
Comment (bendikro) [inline deluge/ui/console/commands/config.py]: Like here `metavar="<value>"`

File: deluge/ui/console/commands/config.py [hunk 4]
```diff
@@ -65,14 +65,14 @@ def simple_eval(source):
...
         get_group = parser.add_argument_group("getting values")
-        get_group.add_argument("keys", metavar="<keys>", nargs="*", help="one or more keys separated by space")
+        get_group.add_argument("keys", metavar="<keys>", nargs="*", help=_("one or more keys separated by space"))
```
Comment (bendikro) [inline deluge/ui/console/commands/config.py]: And `metavar="<keys>"`

File: deluge/ui/console/commands/config.py (hunk 2 above)
Comment (cas--) [inline deluge/ui/console/commands/config.py]: Yes I did ponder that and it's a good point about the `key` and `value`... I'll have a think :)

**Comment by cas--** [comment]
There are the two tickets in 1.3.13 milestone that could also be addressed in this PR. Here is a query of outstanding [console tickets](http://dev.deluge-torrent.org/query?status=accepted&status=assigned&status=new&status=pending&status=reopened&component=Console&milestone=!not+applicable&milestone=!2.0&col=id&col=summary&col=component&col=milestone&col=status&col=type&col=priority&order=priority)

**Comment by cas--** [comment]
Still not working with my [console.conf](https://gist.github.com/cas--/b8c9e3510223bacaf8bb757ff70ef074) and it looks like it because `sidebar_width` is a missed key in my config.

```
  File "/home/ubuntu/delugeDev/deluge-console", line 9, in <module>
    load_entry_point('deluge', 'console_scripts', 'deluge-console')()
  File "/home/ubuntu/deluge.git/deluge/ui/console/__init__.py", line 16, in start
    Console().start()
  File "/home/ubuntu/deluge.git/deluge/ui/console/console.py", line 116, in start
    do_profile=self.options.profile)
  File "/home/ubuntu/deluge.git/deluge/common.py", line 1065, in run_profiled
    return func(*args)
  File "/home/ubuntu/deluge.git/deluge/ui/console/console.py", line 110, in run
    return c.start_ui()
  File "/home/ubuntu/deluge.git/deluge/ui/console/main.py", line 156, in start_ui
    curses.wrapper(self.run)
  File "/usr/lib/python2.7/curses/wrapper.py", line 43, in wrapper
    return func(stdscr, *args, **kwds)
  File "/home/ubuntu/deluge.git/deluge/ui/console/main.py", line 224, in run
    torrentlist = self.register_mode(TorrentList(self.stdscr, self.encoding))
  File "/home/ubuntu/deluge.git/deluge/ui/console/modes/torrentlist/torrentlist.py", line 115, in __init__
    self.sidebar = FilterSidebar(self, self.config)
  File "/home/ubuntu/deluge.git/deluge/ui/console/modes/torrentlist/filtersidebar.py", line 33, in __init__
    width = self.config["torrentview"]["sidebar_width"]
KeyError: 'sidebar_width'
```





If I delete `console.conf` I get this:
```
Traceback (most recent call last):
  File "/home/ubuntu/delugeDev/deluge-console", line 9, in <module>
    load_entry_point('deluge', 'console_scripts', 'deluge-console')()
  File "/home/ubuntu/deluge.git/deluge/ui/console/__init__.py", line 16, in start
    Console().start()
  File "/home/ubuntu/deluge.git/deluge/ui/console/console.py", line 116, in start
    do_profile=self.options.profile)
  File "/home/ubuntu/deluge.git/deluge/common.py", line 1065, in run_profiled
    return func(*args)
  File "/home/ubuntu/deluge.git/deluge/ui/console/console.py", line 110, in run
    return c.start_ui()
  File "/home/ubuntu/deluge.git/deluge/ui/console/main.py", line 156, in start_ui
    curses.wrapper(self.run)
  File "/usr/lib/python2.7/curses/wrapper.py", line 43, in wrapper
    return func(stdscr, *args, **kwds)
  File "/home/ubuntu/deluge.git/deluge/ui/console/main.py", line 218, in run
    self.config.run_converter((0, 1), 2, self._migrate_config_1_to_2)
  File "/home/ubuntu/deluge.git/deluge/config.py", line 519, in run_converter
    raise ex
KeyError: 'sort_primary'
```

**Comment by cas--** [comment]
It loads but see this error...

```
Unhandled Error
Traceback (most recent call last):
File "/home/ubuntu/deluge.git/deluge/transfer.py", line 118, in _handle_complete_message
self.message_received(rencode.loads(zlib.decompress(data), decode_utf8=True))
File "/home/ubuntu/deluge.git/deluge/ui/client.py", line 123, in message_received
d.callback(request[2])
File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 362, in callback
self._startRunCallbacks(result)
File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 458, in _startRunCallbacks
self._runCallbacks()
--- <exception caught here> ---
File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 545, in _runCallbacks
current.result = callback(current.result, *args, **kw)
File "/home/ubuntu/deluge.git/deluge/ui/console/modes/torrentlist/torrentview.py", line 123, in update_state
self.sorted_ids = self._sort_torrents(state)
File "/home/ubuntu/deluge.git/deluge/ui/console/modes/torrentlist/torrentview.py", line 258, in _sort_torrents
if self.config["separate_complete"]:
File "/home/ubuntu/deluge.git/deluge/config.py", line 224, in __getitem__
return self.get_item(key)
File "/home/ubuntu/deluge.git/deluge/config.py", line 245, in get_item
if isinstance(self.__config[key], str):
exceptions.KeyError: 'separate_complete'
```

**Comment by cas--** [comment]
The is a bug when switching filters in sidebar the topmost torrent gets hidden by the column header so you need to scroll up one.

**Comment by cas--** [comment]
I think it would be an improvement to change the selection and popup menu keyshortcuts. 

 * Use `enter` to activate filter in sidebar
 * Use `spacebar` to mark selections e.g. torrents

A few other notes:
* The Deluge version in topbar is unnecessary, creating an 'About' popup would be useful.
 * The 'Current Filter' text in topbar is not well placed currently and a bit superfluous. It would be better placed in the area aboved the sidebar. In fact even better is if the sidebar is hidden and revealed when pressing `left-arrow`. 
* There is no mention of use `spacebar` in help. 
* The help guide is overly verbose. (I can rewrite it if needed)
* Need an option to automatically connect to daemon. 
* While in `help` you cannot open other dialogs straightaway. e.g. preferences
* Need a way to start console in commandline mode.

Bugs:
 * Pressing `h` in certain areas does nothing. e.g. connection manager, sidebar
 * Pressing `o` on a torrent resulted in this error:
```
File "/home/ubuntu/deluge.git/deluge/ui/console/modes/torrentlist/torrentactions.py", line 128, in create_popup          
base_popup=kwargs["base_popup"])
exceptions.KeyError: 'base_popup
```

There are other UI layout changes needing porting over from GTKUI but they can wait.

**Comment by cas--** [comment]
Other than the bugs and anything trivial the above changes can be detailed in a new ticket for future work so this PR can be merged sooner rather than later. I simply thought I'd record my train of thought while it was fresh :)

File: deluge/ui/console/modes/torrentlist/torrentview.py [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+
+        # Just in case primary and secondary fields are empty and/or
+        # both are too ambiguous, also sort by queue position first
+        if "queue" not in [s_secondary, s_primary]:
+            result = sort_by_field(state, result, "queue")
+        if s_secondary != s_primary:
+            result = sort_by_field(state, result, s_secondary)
+        result = sort_by_field(state, result, s_primary)
+
+        if self.config["separate_complete"]:
```
Comment (bendikro) [inline deluge/ui/console/modes/torrentlist/torrentview.py]: Fixed

**Comment by bendikro** [comment]
> Use enter to activate filter in sidebar
> Use spacebar to mark selections e.g. torrents

I don't object to any such changes, however, simply changing to enter in the if test doesn't work, so I didn't look any further at how to solve that. However, it's worth to note that my idea is that, with PR6, it will be possible to have multiple states selected in the filter, which makes sense to use the space key for?
Maybe enter key would select currently marked and deselect any other currently selected?

> There is no mention of use spacebar in help.

No, I've fixed and reworked some the help text, but there is probably much left to be fixed.
> The help guide is overly verbose. (I can rewrite it if needed)

Yeah, I've reduced it somewhat, but it can certainly be refined some more.
> Need an option to automatically connect to daemon.

The help text does say "These daemon connect options will be used for commands, or if console ui autoconnect is enabled." but I haven't seen any _autoconnect_ functionality implemented for this.
> While in help you cannot open other dialogs straightaway. e.g. preferences

True
> Need a way to start console in commandline mode.

Good idea.

> Bugs:
> 
>  *   Pressing h in certain areas does nothing. e.g. connection manager, sidebar

Wouldn't call it bugs. More like non-implemented features :-)
>  *  Pressing o on a torrent resulted in this error:
Fixed

Also fixed bug when switching filter

**Comment by cas--** [comment]
Created new tickets:

Help dialog improvements: [#2916](http://dev.deluge-torrent.org/ticket/2916#ticket)
Sidebar: [#2917](http://dev.deluge-torrent.org/ticket/2917#ticket)
Startup options: [#2918](http://dev.deluge-torrent.org/ticket/2918#ticket)

**Pull request merged**